blacksheep/testing/client.py

```diff
@@ -0,0 +1,64 @@
+from urllib import parse
+
+from blacksheep.messages import Request
+from blacksheep.server.application import Application
+from blacksheep.server.responses import Response
+from blacksheep.testing.helpers import MockSend, get_example_scope
+
+
+def _create_scope(method, path, headers: dict = None, query: dict = b"") -> dict:
+    if headers is not None:
+        headers = [(key.encode(), value.encode()) for key, value in headers.items()]
+
+    if query:
+        query = parse.urlencode(query).encode()
+
+    scope = get_example_scope(method, path, extra_headers=headers, query=query)
+    return scope
+
+
+class _TestSimulator:
+    def __init__(self, app, scope):
+        self.app = app
+        self.scope = scope
+        self._prepare_application()
+
+    async def simulate_request(self) -> Response:
+        request = Request.incoming(
+            self.scope["method"],
+            self.scope["raw_path"],
+            self.scope["query_string"],
+            self.scope["headers"],
+        )
+        response = await self.app.handle(request)
+        return response
+
+    def _prepare_application(self):
+        if self.app._service_provider is None:
+            self.app.build_services()
+            self.app.normalize_handlers()
+            self.app.use_controllers()
+            self.app.configure_middlewares()
+
+
+async def _simulate_request(
+    app: Application, method: str, path: str, headers: dict = None, query: dict = b""
+):
+    scope = _create_scope(method, path, headers, query)
+    simulator = _TestSimulator(app, scope)
+    return await simulator.simulate_request()
+
+
+class TestClient:
+    def __init__(self, app):
+        self.app = app
```

Comment:
I would recommend to instantiate a single `TestSimulator` per TestClient, and maybe to abstract the exact implementation of the object that is responsible of sending web requests. Because then we could support this scenario:

* by default the TestClient would use the `TestSimulator` implementation
* by configuration, TestClient could use a different kind of object to send requests, for example to generate real HTTP requests against a running server, so that the user by configuration can test mocked web requests (enabling super fast tests), or real web requests (useful to test end-to-end integration with real implementations of ASGI servers and http clients)

Something like - high level example -:

````python
from abc import ABC, abstractmethod


class TestHTTPTransport:
    @abstractmethod()
    async def send_request(...) -> void:
        """Sends a web request to the application."""


class TestClient:
    def __init__(self, app, transport: Optional[TestHTTPTransport] = None):
        if transport is None:
             transport = _TestSimulator(app)
        self._transport = transport
    
    async def send_request(...):
        return self._transport.send_request(...)

    async def get(...):
         return await self._send_request(...)
```